src/services/apiService.js

```diff
@@ -0,0 +1,73 @@
+export const apiService = {
+  BASE_API_URL: 'https://blue-bottle-api-test.herokuapp.com',
+  GET_COFFEE_SHOPS: '/v1/coffee_shops',
+  GET_TOKEN: '/v1/tokens',
+
+  async getNearestShops(position) {
+    const coffeeShops = await this.getCoffeeShops();
+    return coffeeShops.map(coffeeShop => {
+      return {
+        id: coffeeShop.id,
+        name: coffeeShop.name,
+        latitude: parseFloat(coffeeShop.x),
+        longitude: parseFloat(coffeeShop.y)
+      }
+    });
+  },
+
+  async getCoffeeShops(apiKey = '') {
+    if (apiKey === '') {
+      apiKey = await this.retrieveNewToken();
+    }
+
+    console.log("Getting the coffee shops...");
+    let response = await fetch(`${this.BASE_API_URL}${this.GET_COFFEE_SHOPS}?token=${apiKey}`);
+
+    if (response.ok) {
+      return await response.json();
+    } else {
+      console.error(`Could not retrieve data. Error code: ${response.status}`);
+
+      if (response.status === 401) {
+        console.log("Getting a new token and trying again...");
+        apiKey = await this.retrieveNewToken();
+
+        if (apiKey !== '') {
```

Comment:
This logic is all good in the case that the apiKey is not valid.
But following the logic from the start of the function until here, I understand that it is not possible to even call the API without the API key, as you always get it from the start. 
Is there a case where this function is called with different api keys, or keys fetched from memory?